walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, or there is no required attendee, 
+        // return meetingTimes for all attendees.
+        if (allMeetingTimes.size() > 0 || requiredAttendees.size() == 0) {
+            return allMeetingTimes;
+        }
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * Pointer rangeStart is pointing at the start of the current available 
+      * TimeRange. If the current available TimeRange has longer duration than the 
+      * requested meeting duration, add it to available meetingTimes.
+      */
+    private static Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events, Collection<String> attendees, long duration) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        // Sort the events by the start time.
+        List<Event> eventsSorted = new ArrayList<Event>(events);
+        Collections.sort(eventsSorted, Event.SORT_BY_START_TIME);
+
+        int rangeStart = DAY_START; // Pointer to start of the previous event.
+
+        // Linear search the current events and add new TimeRange to 
+        // meetingTimes.
+        for (Event event : eventsSorted) {
+            Collection<String> eventAttendees = event.getAttendees();
+            TimeRange eventWhen = event.getWhen();
+            int eventStart = eventWhen.start();
+            int eventEnd = eventWhen.end();
+
+            // Check if given attendees shows up in the event.
+            if (attendees.containsAll(eventAttendees)) {
+                // Check if the current event starts before the previous event ends,
+                // or there is no gap in between.
+                if (rangeStart >= eventStart) {
+                    // If so, check which ends earilier and change the start of next 
+                    // available TimeRange to the later end time of the two.
+                    if (eventEnd > rangeStart) {
+                        rangeStart = eventEnd;
+                    }
+                // If they don't overlap, initialize the current available TimeRange.
+                } else {               
+                    TimeRange newRange = TimeRange.fromStartEnd(
+                            rangeStart, eventStart, false);
+                        if (newRange.duration() >= duration) {
+                            meetingTimes.add(newRange);
+                        }
+                    rangeStart = eventEnd;
+                }
+            }
+        }
+        TimeRange finalRange = TimeRange.fromStartEnd(rangeStart, DAY_END, true);
+        
+        // Check if the duration of the final TimeRange longer than the meeting 
+        // duration.
```

Comment:
This comment doesn't seem to add more information than the line that follows.

Reply:
The comment is removed. :)